src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1138 @@
+
+   //defaults
+   mSetup.processMode = Vst::kOffline;
+   mSetup.symbolicSampleSize = Vst::kSample32;
+   mSetup.maxSamplesPerBlock = 512;
+   mSetup.sampleRate = 44100.0;
+}
+
+VST3Effect::VST3Effect(const VST3Effect& other)
+   : mModule(other.mModule), mEffectClassInfo(other.mEffectClassInfo)
+{
+   Initialize();
+   mSetup = other.mSetup;
+}
+
+VST3Effect::~VST3Effect()
+{
+   using namespace Steinberg;
+
+   CloseUI();
+
+   if(mComponentConnectionProxy)
+      mComponentConnectionProxy->disconnect(FUnknownPtr<Vst::IConnectionPoint>(mEditController));
+   if(mControllerConnectionProxy)
+      mControllerConnectionProxy->disconnect(FUnknownPtr<Vst::IConnectionPoint>(mEffectComponent));
+
+   if(mEditController)
+   {
+      mEditController->setComponentHandler(nullptr);
+      mEditController->terminate();
+   }
+   if(mEffectComponent)
+      mEffectComponent->terminate();
+ 
+   mControllerConnectionProxy = nullptr;
+   mComponentConnectionProxy = nullptr;
+   mAudioProcessor = nullptr;
+   mEditController = nullptr;
+   mEffectComponent = nullptr;
+}
+
+VST3Effect::VST3Effect(
+   std::shared_ptr<VST3::Hosting::Module> module, 
+   VST3::Hosting::ClassInfo effectClassInfo)
+   : mModule(std::move(module)), mEffectClassInfo(std::move(effectClassInfo))
+{
+   using namespace Steinberg;
+   Initialize();
+
+   const auto& pluginFactory = mModule->getFactory();
+   Vst::IEditController* editController;
+   auto result = mEffectComponent->queryInterface(Vst::IEditController::iid, reinterpret_cast<void**>(&editController));
+   if(result != kResultOk)
+   {
+      //Plugin does not have a separate edit controller
+      TUID controllerCID;
+            
+		if (mEffectComponent->getControllerClassId (controllerCID) == kResultTrue)
+		{
+			if(mEditController = pluginFactory.createInstance<Vst::IEditController> (VST3::UID (controllerCID)))
+			{
+            if(mEditController->initialize(LocalContext()) != kResultOk)
+            {
+               wxLogMessage("Cannot create edit controller: initialization failed");
+               mEditController = nullptr;
+            }
+			}
+         else
+            wxLogMessage("Cannot create edit controller: attempt failed");
+		}
+      else
+         wxLogMessage("Cannot create edit controller: failed read controller Class ID from the effect component");
+   }
+   else
+      mEditController = editController;
+
+   if(mEditController)
+   {
+      mComponentHandler = owned(safenew internal::ComponentHandler);
+      mEditController->setComponentHandler(mComponentHandler);
+
+      Steinberg::Vst::IConnectionPoint* componentConnectionPoint { nullptr };
+      Steinberg::Vst::IConnectionPoint* controllerConnectionPoint { nullptr };
+
+      mEffectComponent->queryInterface (Vst::IConnectionPoint::iid, (void**)&componentConnectionPoint);
+      mEditController->queryInterface (Vst::IConnectionPoint::iid, (void**)&controllerConnectionPoint);
```

Comment:
Need two `reinterpret_cast`s for consistent style

Reply:
Turned out there is a safer and shorter for to query interfaces via `FUnknownPtr<>`